### ISSUE TITLE
undefined reference to `pthreadpool_compute_2d_tiled'

### ISSUE THREAD

**Issue by kaishijeng** [facebookarchive/caffe2#459]
When I compile caffe2 using make, I got the following error (ubuntu 16.04):

libCaffe2_CPU.so: undefined reference to `pthreadpool_compute_2d_tiled'

Here is the summary of configuration in my system:

-- ******** Summary ********
-- General:
--   Git version           : v0.7.0-47-g964ffa9-dirty
--   System                : Linux
--   C++ compiler          : /usr/bin/c++
--   Protobuf compiler     : /usr/local/bin/protoc
--   CXX flags             :  -fopenmp -std=c++11 -fPIC -Ofast -ftree-vectorize -ffast-math -funsafe-math-optimizations -ftree-vectorize -fomit-frame-pointer -Wno-narrowing
--   Build type            : Release
--   Compile definitions   : CAFFE2_USE_EIGEN_FOR_BLAS;CAFFE2_USE_GOOGLE_GLOG;CAFFE2_USE_GFLAGS;EIGEN_MPL2_ONLY;CAFFE2_FORCE_FALLBACK_CUDA_MPI;CAFFE2_NO_BUILTIN_CPU_SUPPORTS
--
--   BUILD_SHARED_LIBS     : ON
--   BUILD_PYTHON          : ON
--     Python version      : 2.7.12
--     Python library      : /usr/lib/aarch64-linux-gnu/libpython2.7.so
--   USE_CUDA              : OFF
--   USE_NERVANA_GPU       : OFF
--   USE_GLOG              : ON
--   USE_GFLAGS            : ON
--   USE_LMDB              : ON
--     LMDB version        : 0.9.17
--   USE_LEVELDB           : ON
--     LevelDB version     : 1.18
--     Snappy version      : 1.1.3
--   USE_OPENCV            : OFF
--   USE_ZMQ               : OFF
--   USE_ROCKSDB           : OFF
--   USE_MPI               : ON
--   USE_NCCL              : OFF
--   USE_NNPACK            : ON
--   USE_OPENMP            : ON
--   USE_REDIS             : OFF
--   USE_GLOO              : OFF
-- Configuring done

Any idea of how to resolve it?

Thanks,

**Comment by dbermond** [comment]
I'm having a similar error on Arch Linux x86_64 when using nnpack from system libraries (external nnpack installation):
```
libCaffe2_CPU.so: undefined reference to `pthreadpool_compute_1d_tiled'
libCaffe2_CPU.so: undefined reference to `pthreadpool_compute_2d_tiled'
libCaffe2_CPU.so: undefined reference to `pthreadpool_compute_2d'
libCaffe2_CPU.so: undefined reference to `pthreadpool_create'
collect2: error: ld returned 1 exit status
```

My external (system) nnpack and pthreadpool are compiled with `-fPIC` in `cflags` and `cxxflags`.

Interestingly though, when uninstalling nnpack and letting caffe2 to build nnpack from source during the compilation, I have no error.

**Comment by zfc929** [comment]
@kaishijeng @dbermond 

I modify Depencies.cmake
add :+1: list(APPEND NNPACK_INCLUDE_DIRS "/usr/local/include")
list(APPEND NNPACK_LIBRARIES "/usr/local/lib")
#set(NNPACK_LIBRARIES /usr/local/lib/libnnpack.a /usr/local/lib/libpthreadpool.a)
link_libraries("/usr/local/lib/libnnpack.a" "/usr/local/lib/libpthreadpool.a")

this is ok

**Comment by dbermond** [comment]
@zfc929 Thank you for sharing this solution. It works.

The relevant part though is setting the `NNPACK_LIBRARIES` cmake internal variable. It will define which nnpack libraries will be used when linking: 
[https://github.com/caffe2/caffe2/blob/05ba3e954cd86b7c1cfbb001a675a46412948adb/cmake/Dependencies.cmake#L56](https://github.com/caffe2/caffe2/blob/05ba3e954cd86b7c1cfbb001a675a46412948adb/cmake/Dependencies.cmake#L56)

libpthreadpool is needed during caffe2 linking stage. This can be confirmed by taking a look at the caffe2 building process when the user chooses to build nnpack from source: 
[https://github.com/caffe2/caffe2/blob/05ba3e954cd86b7c1cfbb001a675a46412948adb/cmake/External/nnpack.cmake#L151](https://github.com/caffe2/caffe2/blob/05ba3e954cd86b7c1cfbb001a675a46412948adb/cmake/External/nnpack.cmake#L151)

This libpthreadpool linking part is **missing** from cmake when the user chooses to use an external nnpack installation (from system libraries): 
[https://github.com/caffe2/caffe2/blob/05ba3e954cd86b7c1cfbb001a675a46412948adb/cmake/Modules/FindNNPACK.cmake#L28](https://github.com/caffe2/caffe2/blob/05ba3e954cd86b7c1cfbb001a675a46412948adb/cmake/Modules/FindNNPACK.cmake#L28)

**Comment by dbermond** [comment]
@kaishijeng 
This issue was fixed by commit f070eb99a4a36f12abbdb9af98bb68ff485c840b. At least for me it's fixed.
I think it can be closed.

**Comment by pietern** [comment]
Thanks for confirming @dbermond.
